compiler/optimizer/abstractinterpreter/InliningProposal.cpp

```diff
@@ -0,0 +1,260 @@
+         sprintf(line, "#Proposal: #%d : #%d %s @%d -> bcsz=%d %s target %s, benefit = %f, cost = %d, budget = %d",
+            currentNode->getGlobalIndex(),
+            currentNode->getParentGlobalIndex(),
+            _nodes->isSet(index + 1) ? "INLINED" : "NOT inlined",
+            currentNode->getByteCodeIndex(),
+            currentNode->getByteCodeSize(),
+            currentNode->getResolvedMethodSymbol()->signature(comp->trMemory()),
+            currentNode->getName(comp->trMemory()),
+            currentNode->getBenefit(),
+            currentNode->getCost(),
+            currentNode->getBudget()
+         );
+
+         if (traceBIProposal)
+            traceMsg(comp, "%s\n",line);
+         if (verboseInlining)
+            TR_VerboseLog::writeLineLocked(TR_Vlog_BI, line);
+         }
+
+      //process children
+      const uint32_t numChildren = currentNode->getNumChildren();
+
+      for (uint32_t i = 0; i < numChildren; i++)
+         idtNodeQueue.push_back(currentNode->getChild(i));
+
+      }
+
+   traceMsg(comp, "\n");
+   }
+
+
+void TR::InliningProposal::addNode(TR::IDTNode *node)
+   {
+   ensureBitVectorInitialized();
+
+   const int32_t index = node->getGlobalIndex() + 1;
+   if (_nodes->isSet(index))
+      {
+      return;
+      }
+
+   _nodes->set(index);
+
+   _cost = 0;
+   _benefit = 0;
+   }
+
+bool TR::InliningProposal::isEmpty()
+   {
+   if (!_nodes)
+      return true;
+
+   return _nodes->isEmpty();
+   }
+
+uint32_t TR::InliningProposal::getCost()
+   {
+   if (_cost == 0)
+      {
+      computeCostAndBenefit();
+      }
+
+   return _cost;
+   }
+
+double TR::InliningProposal::getBenefit()
+   {
+   if (_benefit == 0)
+      {
+      computeCostAndBenefit();
+      }
+
+   return _benefit;
+   }
+
+void TR::InliningProposal::computeCostAndBenefit()
+   {
+   _cost = 0;
+   _benefit = 0;
+
+   if (!_idt)
+      return;
+
+   TR_BitVectorIterator bvi(*_nodes);
+   int32_t idtNodeIndex;
+
+   while (bvi.hasMoreElements())
+      {
+      idtNodeIndex = bvi.getNextElement();
+      IDTNode *node = _idt->getNodeByGlobalIndex(idtNodeIndex - 1);
+      if (node == NULL)
+         {
+         continue;
+         }
+      _cost += node->getCost();
+      _benefit += node->getBenefit();
+      }
+   }
+
+void TR::InliningProposal::ensureBitVectorInitialized()
+   {
+   if (!_nodes)
+      _nodes = new (_region) TR_BitVector(_region);
+   }
+
+bool TR::InliningProposal::isNodeInProposal(TR::IDTNode* node)
+   {
+   if (node == NULL)
+      return false;
+   if (_nodes == NULL)
+      return false;
+   if (_nodes->isEmpty())
+      return false;
+
+   const int32_t idx = node->getGlobalIndex() + 1;
+
+   return _nodes->isSet(idx);
+   }
+
+void TR::InliningProposal::unionInPlace(TR::InliningProposal *a, TR::InliningProposal *b)
+   {
+   ensureBitVectorInitialized();
+   a->ensureBitVectorInitialized();
+   b->ensureBitVectorInitialized();
+
+   *_nodes = *a->_nodes;
+   *_nodes |= *b->_nodes;
```

Comment:
I amended this part by using an additional temporary variable to make it flawless under the situation of `b` aliasing `this`. Since this not a in-place union anymore, I also changed the signature of this method to `merge()`.